src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/HtmlFormattingPass.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Formatting
+{
+    internal class HtmlFormattingPass : FormattingPassBase
+    {
+        private readonly ILogger _logger;
+
+        public HtmlFormattingPass(
+            RazorDocumentMappingService documentMappingService,
+            FilePathNormalizer filePathNormalizer,
+            ILanguageServer server,
+            ILoggerFactory loggerFactory)
+            : base(documentMappingService, filePathNormalizer, server)
+        {
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<HtmlFormattingPass>();
+        }
+
+        // We want this to run first because it uses the client HTML formatter.
+        public override int Order => DefaultOrder - 5;
+
+        public async override Task<FormattingResult> ExecuteAsync(FormattingContext context, FormattingResult result)
+        {
+            if (context.IsFormatOnType)
+            {
+                // We don't want to handle OnTypeFormatting here.
+                return result;
+            }
+
+            var originalText = context.CodeDocument.GetSourceText();
+
+            var htmlEdits = await HtmlFormatter.FormatAsync(context.CodeDocument, context.Range, context.Uri, context.Options);
```

Comment:
So currently we ask the client to format correct?

Reply:
Yes